src/transformers/modeling_bart.py

```diff
@@ -0,0 +1,1062 @@
+
+
+# Public API
+
+
+@add_start_docstrings(
+    "The bare BART Model model outputting raw hidden-states without any specific head on top.",
+    BART_START_DOCSTRING,
+    BART_INPUTS_DOCSTRING,
+)
+class BartModel(PretrainedBartModel):
+    """"""
+
+    def __init__(self, config: BartConfig):
+        super().__init__(config)
+        self.output_attentions = config.output_attentions
+        self.output_hidden_states = config.output_hidden_states
+
+        padding_idx, vocab_size = config.pad_token_id, config.vocab_size
+        self.shared = nn.Embedding(vocab_size, config.d_model, padding_idx)
+
+        self.encoder = BartEncoder(config, self.shared)
+        self.decoder = BartDecoder(config, self.shared)
+
+        self.init_weights()
+
+    def get_input_embeddings(self):
+        return self.shared
+
+    def set_input_embeddings(self, value):
+        self.shared = value
+
+    def get_output_embeddings(self):
+        return _make_linear_from_emb(self.shared)
+
+    def forward(
+        self,
+        input_ids,
+        attention_mask=None,
+        decoder_input_ids=None,
+        encoder_outputs=None,  # type: Tuple
+        decoder_attention_mask=None,
+        cached_states=None,
+        **unused  # TODO(SS): does fairseq have an equivalent to token_type_ids?
+    ):
+        """
+        Args:
+            input_ids:
+            attention_mask: Ignore pad tokens in the input_ids
+            decoder_input_ids:
+            encoder_outputs: None or Tuple. If a tuple is passed, its first entry is assumed to be the features. If
+            none is passed, we call the encoder.
+            decoder_attention_mask:
+            **unused:
+
+        Returns:
+
+        """
+        # make masks if user doesn't supply
+        attention_mask, decoder_input_ids, decoder_attn_mask = prepare_bart_inputs(
+            self.config,
+            input_ids,
+            attention_mask=attention_mask,
+            decoder_input_ids=decoder_input_ids,
+            decoder_attn_mask=decoder_attention_mask,
+        )
+        assert decoder_input_ids is not None
+        if encoder_outputs is None:
+            # TODO(SS): make this caching more usable when overwrite generate
+            encoder_outputs = self.encoder.forward(input_ids=input_ids, attention_mask=attention_mask)
+        assert isinstance(encoder_outputs, tuple)
+        # dec_features, cached_states, dec_hidden, dec_attn
+        decoder_outputs = self.decoder.forward(
+            decoder_input_ids, encoder_outputs[0], attention_mask, decoder_attention_mask, cached_states=cached_states,
+        )
+        # Attention and hidden_states will be [] or None if they aren't needed
+        decoder_outputs = _filter_out_falsey_values(decoder_outputs)  # type: tuple
+        assert isinstance(decoder_outputs[0], torch.Tensor)
+        encoder_outputs = _filter_out_falsey_values(encoder_outputs)  # type: tuple
+        return decoder_outputs + encoder_outputs
```

Comment:
for language generation we would need the following variables from `decoder_outputs + encoder_outputs`:
the variable `output` from  `decoder_outputs[0]`, which will be transformed to logits be the `lm_head` in `BartForMaskedLM`, the variable `cached_states` from `decoder_outputs[2]` for faster decoding, the variable `encoder_hidden_states` from `encoder_outputs[0]` and the variable `inputs` which shouldn't change and is only needed for the first decoding step and afterwards by the `prepare_bart_inputs` fn.